### PR TITLE
#161348792 Users should receive descriptive signup/Registration validation errors

### PR DESCRIPTION
**Pull request by dnuwa**
#### What does this PR do?
Return descriptive error message to the user
#### Description of Task to be completed?
Have inputs validated on the `POST /api/users/` endpoint ensuring that descriptive error messages are returned to the user
##### To register a user
- Run the application with postman and attempt registering user with incorrect data.
url: api/users
payload:
```
{"user":
	{
		"email":"testuser@text.com",
		"username":"testuser",
		"password":"userpassword"
	
	}	
}
```
##### Error returned
```
{
    "errors": {
        "error": [
            "Ensure your password is alphanumeric, with Minimum eight characters, at least one letter, one number and one special character"
        ]
    }
}
```
#### Any background context you want to provide?
When a user attempts registering with a password that is not alphanumeric and no other special character, the API returns a descriptive error message
This also applies to attempts on submitting an empty password, email, and username fields
#### What are the relevant pivotal tracker stories?
[#161348792](https://www.pivotaltracker.com/story/show/161348792)

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -4,26 +4,51 @@
...
-        min_length=8,
-        write_only=True
-    )
+    password = serializers.CharField(max_length=128, min_length=8, write_only=True)
 
     # The client should not be able to send a token along with a registration
     # request. Making `token` read-only handles that for us.
 
+    def reg_fields_validation(self, username, email, password):
+        # if email is None:
```
Comment (emugaya) [inline authors/apps/authentication/serializers.py]: Remove commented code

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -4,26 +4,51 @@
...
+        #     raise serializers.ValidationError("An email address is required")
+
+        if re.compile("[!@#$%^&*:;?><.0-9]").match(username):
+            raise serializers.ValidationError(
+                "Invalid Username , it contains invalid characters."
+            )
+
+        if str(password).isdigit() :
```
Comment (emugaya) [inline authors/apps/authentication/serializers.py]: use one regex for both alphanumeric e.g `^(?=.*[a-z])(?=.*\d)(?=.*[A-Z])([a-zA-Z\d]{8})$` pattern can take care of your password validations instead of multiple if statements.

File: authors/apps/authentication/tests/test_model.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+from rest_framework.test import APITestCase
```
Comment (emugaya) [inline authors/apps/authentication/tests/test_model.py]: Model testing is good but let's focus on testing endpoints. Please organize tests in folders relating to what your building e.g `auth`

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -4,26 +4,51 @@
...
+
+        self.reg_fields_validation(username, email, password)
+        return {"email": email, "username": username, "password": password}
+
     class Meta:
         model = User
         # List all of the fields that could possibly be included in a request
         # or response, including fields specified explicitly above.
-        fields = ['email', 'username', 'password']
+        fields = ["email", "username", "password"]
```
Comment (judeinno) [inline authors/apps/authentication/serializers.py]: Why are you changing from single quotes to double quotes? Because am using some of those fields and it may lead to merge conflicts.

**Comment by brucemakallan** [comment]
We agreed on a tearDown() as a team to avoid creating a new user every time we are testing. I wasn't able to find it in the changed files. Did you find an alternative?

File: authors/apps/authentication/urls.py [hunk 5]
```diff
@@ -6,6 +6,6 @@
 
 urlpatterns = [
```
Comment (Sulaiman-Mozes) [inline authors/apps/authentication/urls.py]: Its better to provide names for all the urls than just one

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (RachaelNantale) [inline authors/apps/authentication/serializers.py]: Why did you change to double quotes?

File: authors/apps/authentication/urls.py (hunk 5 above)
Comment (dnuwa) [reply authors/apps/authentication/urls.py]: @Sulaiman-Mozes kindly color more on this

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (dnuwa) [reply authors/apps/authentication/serializers.py]: @RachaelNantale this was done by pep8 formatting. since we resolved to use single quotes, this file has been updated to have single quotes.

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (dnuwa) [reply authors/apps/authentication/serializers.py]: Done and it turns out it's very effective.

**Comment by emugaya** [comment]
@dnuwa Please follow conventions in the engineering playbook on how to write PRs

File: authors/apps/authentication/validators.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    def is_username_valid(self, username):
+        if len(username) < 4:
+            raise APIException(
+                {'username_error': 'The username should be atleast 4 characters'}
+            )
+        if re.compile('[!@#$%^&*:;?><.0-9]').search(username):
+            raise APIException(
+                {'username_error':'Invalid Username , it contains invalid characters.'}
+            )
+
```
Comment (emugaya) [inline authors/apps/authentication/validators.py]: Add newline

**Review by emugaya — APPROVED** [review]
Fix Heroku review App before I merge

File: authors/apps/authentication/tests_auth/test_model.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        )
+
+    def test_create_single_user(self):
+        user_count = User.objects.count()
+        self.assertEqual(user_count, 2)
+        self.assertEqual(self.user.username, "testuser")
+        self.assertEqual(self.user.email, "test@test.com")
+
+    def test_missing_username(self):
+        try:
```
Comment (emugaya) [inline authors/apps/authentication/tests_auth/test_model.py]: Why do you have `try catch` blocks in tests?

**Review by emugaya — CHANGES_REQUESTED** [review]
Your model tests are lacking. Read on how to write `django` model tests. I advised against writing them unless your sure of what you are doing.

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (emugaya) [reply authors/apps/authentication/serializers.py]: @dnuwa Please respond to this comment

**Review by RachaelNantale — APPROVED** [review]

**Review by judeinno — APPROVED** [review]

**Review by emugaya — APPROVED** [review]
